Userland/Libraries/LibCompress/Deflate.cpp

```diff
@@ -90,14 +84,17 @@ Optional<CanonicalCode> CanonicalCode::from_bytes(ReadonlyBytes bytes)
 
             code.m_symbol_codes.append(start_bit | next_code);
             code.m_symbol_values.append(symbol);
+            code.m_bit_codes[symbol] = fast_reverse16(start_bit | next_code, code_length); // DEFLATE writes huffman encoded symbols as lsb-first
+            code.m_bit_code_lengths[symbol] = code_length;
 
             next_code++;
         }
     }
 
-    if (next_code != (1 << 15)) {
-        return {};
-    }
+    // FIXME: I'm not quite sure what this is checking for, but it fails for certain (seemingly) valid codes
+    // if (next_code != (1 << 15)) {
+    //     return {};
+    // }
```

Comment:
> FIXME: I'm not quite sure what this is checking for, but it fails for certain (seemingly) valid codes

Me neither. The [tinflate.c](https://achurch.org/tinflate.c) DEFLATE implementation is simplistic and makes for a good comparison.

It has a similar code pattern including this check in `gen_huffman_table`.

```
    if (first_code[15] + length_count[15] != (unsigned short)1<<15) {
        /* The Huffman tree is incomplete and thus invalid. */
        return 0;
    }
```

The calculation of `first_code` (`next_code` in Serenity's implementation) is also different.

```
first_code[i] = (first_code[i-1] + length_count[i-1]) << 1;
```

Reply:
While i'm still not entirely sure what that checks for, that library has a special case for canonical codes with exactly 1 symbol, and in fact the valid codes which this check failed for are codes with exactly 1 symbol, so ill just add a similar special case for 1 symbol and re-enable the check :)